### ISSUE TITLE
JSON-RPC for driving symbolic execution

### ISSUE THREAD

**Issue by goodlyrottenapple** [runtimeverification/haskell-backend#3061]
This is a proposal to switch to using the JSON RPC protocol to talk with kore-repl/kore-exec. The main proposed change would be to remove the current REPL and replace it with a JSON RPC server which would handle the "core" functionality of the current REPL, namely all the functionality such as step/stepf, config, select, etc. excluding any complex commands like piping or appending to file, which would instead be delegated to a new REPL. This new REPL could be written in Haskell or any potentially any other language, e.g. Python, and would simply be a JSON RPC client together with the REPL parser pretty printer, etc. This REPL could potentially live in the K framework repo and thus have better access to higher level functionality such as the pretty printing of kore ASTs back into K (since `config` in the current REPL returns a kore AST). A potential advantage of this approach is a common interface for both a REPL and the summarizer to interact with the haskell backend. 

Propsed features
* new REPL should mirror the current REPL functionality
* have built in pretty printing of K terms (discussed in #1890, #2797)
* should be interruptible/cancellable  #1691
* have state diffing built in

Questions
* What language to use for the new REPL and which repo to put it in?
* Could the LLVM backend also have a JSON RPC interface? Would this make sense?

### Edit: 12 July 2022
We met and discussed that we need the following:

- [x] #3143
- [x] #3144
- [x] #3148 
- [x] #3146

The step request will just be a particular case of the execute request. The execute request will also return the branching states etc. together with the parent node of the branch.

The check-implication request can just expose the current check implication procedure we have, with some minimal changes.

**Comment by goodlyrottenapple** [comment]
# Implementation details

We could use [servant-jsonrpc](https://hackage.haskell.org/package/servant-jsonrpc-1.1.1/docs/Servant-JsonRpc.html) to pull out the current functionality of the REPL into the following API

```haskell
type Step   = JsonRpc "step"   Int Text [Int] 
```
_The `Step` method expects an Int argument and either returns a `Text` error or a list of new states in case of branching. Maybe instead of `[Int]` we should return a hash of the state as it appears in the summarizer?? Or return both?_

```haskell
type Stepf  = JsonRpc "stepf"  Int Text [Int] 
type Select = JsonRpc "select" Int Text ()
```
_Should `Select` take the hash instead? Or maybe have two selects, one with an int and the other with a hash?_
```haskell
type Config = JsonRpc "config" Int Text Text
```
_Not sure whether `Config` should return a structured type or just dump the kore state as text._

```haskell
type Cancel = JsonRpcNotification "cancel" ()
```
_Should cancel the current operation, most likely to be used with `Step(f)`_

```haskell
type RpcAPI = Step :<|> Stepf :<|> Select :<|> Config :<|> Cancel
```

This RPC server would essentially be the same stateful interpreter that we currently have, as we want to avoid the startup cost of reloading the current file for each individual command, as is currently the case for the KSummarizer. However, since we want to add a cancel operation, we would probably? have to tweak the code to be able to handle the `Cancel`. Probably best to implement the server to run two threads, the worker thread that does one computation at a time and the server thread that buffers the messages and sends back responses.

**Comment by goodlyrottenapple** [comment]
Actually I found a different library https://hackage.haskell.org/package/json-rpc-1.0.4/docs/Network-JSONRPC.html which may be more suited if we want to support bi-directional communication, e.g. when having debugging enabled. The servant-jsonrpc library assumes a client-server architecture, where the client always initiates a request and then waits for the response. However, we might want to add an RPC function to e.g. enable debugging or tracing information. Then, instead of the client waiting on a response to an "enable debugging info" message, the server should instead start sending individual debug messages to the client whenever the execution of e.g. step calls for it. This would change the server into a sort of pub-sub configuration rather than a straightforward client-server. I believe the json-rpc library mentioned above supports both models.... @ana-pantilie @ehildenb is this something you can see we might want to support?

**Comment by goodlyrottenapple** [comment]
Here is the proposed API implementation agnostic view:

## Messages from client to the kore-exec JSON RPC server
| Method name     | Parameters   | Return type    | Message type  | Notes             |
| ---------------- | ------------- | ------------- | --------------- | ------------- |
| step                     | Int                  | [Int]                | client-server | The `Step` method expects an Int argument and either returns a `Text` error or a list of new states in case of branching. Maybe instead of `[Int]` we should return a hash of the state as it appears in the summarizer?? Or return both? |
| stepf                    | Int                  | [Int]                 | regular          |  |
| select                   | Int                  | ()                     | regular          | Should `Select` take the hash instead? Or maybe have two selects, one with an int and the other with a hash? |
| config                   | Int                  | Text                | regular         | Not sure whether `Config` should return a structured type or just dump the kore state as text.|
| cancel                  |                        |                       | notification   | Should cancel the current operation, most likely to be used with `step(f)` |
| log_subscribe     | [Text]             |                       | notification (pub-sub) | Subscribe to the log messages passed in via the parameter, e.g. `["DebugAttemptEquation", "DebugApplyEquation"]` |
| log_unsubscribe | [Text]             |                       | notification (pub-sub) | opposite of `og_subscribe`  |

## Messages from the kore-exec JSON RPC server to the client
| Method name     | Parameters   | Return type    | Message type | Notes
| ---------------- | ------------- | ------------- | --------------- | ------------- |
| log_message      | LogType        |                        | notification       | `LogType` should look something like `{"type": "DebugAttemptEquation", "message": "..." }` |


## Message type explanation

| Message type  | Description             |
| --------------  | ------------------ | 
| regular              | Regular JSON RPC request message, where the client should expect a response. See https://www.jsonrpc.org/specification#request_object | 
| notification       | JSON RPC notification message, where the receiver does not send a response back. See https://www.jsonrpc.org/specification#request_object |
| pub-sub            | A regular or notification JSON RPC message from the subscriber to the publisher, which may result in the publisher sending notification messages back to the subscriber. E.g. see https://geth.ethereum.org/docs/rpc/pubsub |

**Comment by ehildenb** [comment]
rule <k> X:Int => a ... </k> requires X <Int 3

```
call: {"step": "<k> Y:Int </k>"}
result: [
{ term: "<k> Y </k>", "condition": "notBool X <Int 3", "depth": 0 },
{ term: "<k> a </k>", "condition": "X <Int 3", "depth": 1 }
]
```

**Comment by ehildenb** [comment]
Let's make the architecture so that there is one main server thread which accepts/sends requests. Add a cancel command which can cancel a given command running.

**Comment by ehildenb** [comment]
Optional bonus that makes things way easier later:

```
{ "add-axiom": KORE_AXIOM }
```

The added axiom should be able to be injected at any priority. The axiom should be able to be a normal semantic `rule`, or a `simplification` rule.

The response should just be "true" for "it worked"

**Comment by ana-pantilie** [comment]
### Server start-up

The configuration looks like:
```
data KoreRpcServerOptions =
   KoreRpcServerOptions
       { definitionFileName :: FilePath
        , mainModuleName :: ModuleName
        , koreSolverOptions :: KoreSolverOptions
        , koreLogOptions :: KoreLogOptions
        , version :: _
       }
```

Start-up will load and initialize the definition, meaning it will basically do the same thing as https://github.com/runtimeverification/haskell-backend/blob/22b637981b4b5ff86963fe9fbc8322e7b65b26ed/kore/app/exec/Main.hs#L662 until line 665. This only does syntactical validation of the kore definition but it doesn't internalize the definition. What this means is that lines 668-670 take care of extracting the necessary info from the definition (axioms and stuff) and we need to separate that from actual the code doing the actual execution in order to implement server start-up in a reasonable way.

**Comment by ana-pantilie** [comment]
Next steps:
- review https://github.com/runtimeverification/haskell-backend/pull/3068
- see if it's possible to restrict to one session
- initialize server by spawning SMT process and initializing and internalizing the definition, basically doing everything up until https://github.com/runtimeverification/haskell-backend/blob/dea085939a1c1005fb3216f7e7a0890c3e34a3a6/kore/app/exec/Main.hs#L695 and using the simplifier init function from above PR

**Comment by goodlyrottenapple** [comment]
from @nishantjr:

We noticed while looking at the logs, that when executing up to a branch point, the backend may be calculating the states immediately after the branch point as well. If that's the case, could the JSON RPC also return those states?
We'd like to change the result of the `execute` action from:
```
"state": KORE_TERM
"reason": REASON_FOR_HALTING
```
Where `KORE_TERM` is the final term, and `REASON_FOR_HALTING` is one of:

```
{ "reason": "branching", "depth": DEPTH }
{ "reason": "stuck", "depth": DEPTH }
{ "reason": "depth-bound" }
{ "reason": "pattern-match", "depth": DEPTH, "matches": [ { "pattern": INT_INDEX_OF_PATTERN_THAT_MATCHED, "substitution": MATCHING_SUBSTITUTION } ] }
```
to:
```
"state": KORE_TERM
"reason": REASON_FOR_HALTING
```
Where `KORE_TERM` is the final term, and `REASON_FOR_HALTING` is one of:

```
{ "reason": "branching", "depth": DEPTH, "post-branch-states": [ { "state": POST_BRANCH_STATE, "depth": POST_BRANCH_DEPTH, ...]}
{ "reason": "stuck", "depth": DEPTH }
{ "reason": "depth-bound" }
{ "reason": "pattern-match", "depth": DEPTH, "matches": [ { "pattern": INT_INDEX_OF_PATTERN_THAT_MATCHED, "substitution": MATCHING_SUBSTITUTION } ] }
```

**Comment by ehildenb** [comment]
No need for frontend or LLVM backend support here, we already have good textual/binary format for frontend/backend talknig to each other.

Maybe later we need it, and we can add it then. Currently we only need pyk/haskell-backend to talk to each other.

**Comment by ehildenb** [comment]
We actually don't need support for `patterns` in `execute-to-branch` anymore, instead we can support `cut-point-rules` and `terminal-rules`.

`cut-point-rules` is a set of string rule labels/identifiers (like `rule [my-rule-name]: ....`), which if one of these rules applies, we want the state _before_ the rule applied to be returned. `terminal-rules` is a set of string rule labels/identifiers (like `rule [my-rule-name]: ...`) which if one of these rules applies, we want the state _after_ the rule applied to be returned.

**Comment by ana-pantilie** [comment]
Meeting discussion with @goodlyrottenapple:
- check implication should return true/false and an optional substitution and predicate
- return dummy responses for each command
- prioritize wrapping up the API

Future work:
- https://github.com/runtimeverification/haskell-backend/issues/3148 is highest priority after getting the API merged
- discuss with Pyk team about giving us examples

**Comment by ana-pantilie** [comment]
@goodlyrottenapple just export types from the Kore JSON module.

**Comment by ana-pantilie** [comment]
Next step as of today: additional testing, and look at `check-implication`.
